File: go/decimal.go
/* Copyright (c) 2025 Richard Rodger, MIT License */

package aontu

import (
	"math/big"
	"strings"
)


type Decimal struct {
	coeff *big.Int
	scale int32
}

const (
	decimalMaxCoeffDigits = 4096
	decimalMaxScale       = 4096
)

func newDecimal(coeff *big.Int, scale int32) *Decimal {
	c := new(big.Int).Set(coeff)

	if c.Sign() == 0 {
		return &Decimal{coeff: c, scale: 1}
	}

	// Strip trailing zeros, but never below one decimal place.
	ten := big.NewInt(10)
	q, m := new(big.Int), new(big.Int)
	for scale > 1 {
		q.QuoRem(c, ten, m)
		if m.Sign() != 0 {
			break
		}
		c.Set(q)
		scale--
	}

	if scale < 1 {
		c.Mul(c, new(big.Int).Exp(ten, big.NewInt(int64(1-scale)), nil))
		scale = 1
	}

	return &Decimal{coeff: c, scale: scale}
}

func (d *Decimal) neg() *Decimal {
	if d.coeff.Sign() == 0 {
		return d
	}
	return &Decimal{coeff: new(big.Int).Neg(d.coeff), scale: d.scale}
}

func (d *Decimal) add(o *Decimal) *Decimal {
	a, b := d.coeff, o.coeff
	scale := d.scale
	switch {
	case d.scale < o.scale:
		a = shiftUp(a, int64(o.scale)-int64(d.scale))
		scale = o.scale
	case o.scale < d.scale:
		b = shiftUp(b, int64(d.scale)-int64(o.scale))
	}
	return newDecimal(new(big.Int).Add(a, b), scale)
}

func (d *Decimal) mul(o *Decimal) *Decimal {
	return newDecimal(new(big.Int).Mul(d.coeff, o.coeff),
		d.scale+o.scale)
}

func (d *Decimal) ceilFloor(up bool) *Decimal {
	// The normal form guarantees scale >= 1, so the divisor is >= 10.
	q, r := new(big.Int), new(big.Int)
	q.QuoRem(d.coeff, pow10(int64(d.scale)), r)
	switch {
	case up && r.Sign() > 0:
		q.Add(q, big.NewInt(1))
	case !up && r.Sign() < 0:
		q.Sub(q, big.NewInt(1))
	}
	return newDecimal(q, 0)
}

func (d *Decimal) overBudget() bool {
	if decimalMaxScale < d.scale || d.scale < -decimalMaxScale {
		return true
	}
	return decimalMaxCoeffDigits < len(new(big.Int).Abs(d.coeff).String())
}

func (d *Decimal) cmp(o *Decimal) int {
	if d.scale == o.scale {
		return d.coeff.Cmp(o.coeff)
	}
	a, b := d.coeff, o.coeff
	if d.scale < o.scale {
		a = shiftUp(a, int64(o.scale)-int64(d.scale))
	} else {
		b = shiftUp(b, int64(d.scale)-int64(o.scale))
	}
	return a.Cmp(b)
}

// equal reports exact numeric equality.
func (d *Decimal) equal(o *Decimal) bool { return d.cmp(o) == 0 }

// shiftUp returns n * 10^by (by >= 0), leaving n untouched.
func shiftUp(n *big.Int, by int64) *big.Int {
	return new(big.Int).Mul(n, pow10(by))
}

// pow10 returns 10^n as an exact integer (n >= 0).
func pow10(n int64) *big.Int {
	return new(big.Int).Exp(big.NewInt(10), big.NewInt(n), nil)
}

func (d *Decimal) digits() string {
	ds := new(big.Int).Abs(d.coeff).String()
	sc := int(d.scale)
	// The normal form guarantees sc >= 1; pad so there is always at
	// least one digit before the point ("0.1", not ".1").
	if len(ds) <= sc {
		ds = strings.Repeat("0", sc-len(ds)+1) + ds
	}
	out := ds[:len(ds)-sc] + "." + ds[len(ds)-sc:]
	if d.coeff.Sign() < 0 {
		out = "-" + out
	}
	return out
}

// String renders the decimal as its plain digits (no `0d` marker), so a
// Decimal formats readably in Go error and debug output.
func (d *Decimal) String() string { return d.digits() }

func (d *Decimal) MarshalJSON() ([]byte, error) { return []byte(d.digits()), nil }

// Canon renders the decimal as an Aontu literal: the sign BEFORE the
// marker (`-0d1.5`), plain form at every magnitude (D4).
func (d *Decimal) Canon() string { return markExact(d.digits()) }

func markExact(digits string) string {
	if strings.HasPrefix(digits, "-") {
		return "-0d" + digits[1:]
	}
	return "0d" + digits
}

// bigIntDigits renders a biginteger peg as plain digits (no marker).
func bigIntDigits(n *big.Int) string { return n.String() }

File: ts/src/val/Decimal.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */


const DECIMAL_COEFFICIENT_BUDGET = 4096
const DECIMAL_SCALE_BUDGET = 4096


function pow10(n: number): bigint {
  return 10n ** BigInt(n)
}


function overBudget(coeffDigits: number, scale: number): boolean {
  return DECIMAL_COEFFICIENT_BUDGET < coeffDigits ||
    !(Math.abs(scale) <= DECIMAL_SCALE_BUDGET)
}


// The decimal digit count of a coefficient, sign ignored (zero counts as
// one digit). Only the budget test needs it.
function coeffDigits(unscaled: bigint): number {
  return (unscaled < 0n ? -unscaled : unscaled).toString().length
}


class Decimal {
  readonly unscaled: bigint

  // Always >= 1: an integral bigdecimal keeps one decimal place so its
  // canon cannot reparse as a biginteger.
  readonly scale: number

  constructor(unscaled: bigint, scale: number) {
    let u = unscaled
    let s = Math.trunc(scale)

    if (0n === u) {
      s = 1
    }
    else {
      // Strip the trailing zeros that scale-as-presentation would
      // otherwise smuggle into identity (`0d0.10` -> `0d0.1`)...
      while (1 < s && 0n === u % 10n) {
        u /= 10n
        s--
      }
      // ...and fold a negative scale (an exponent) back up to the one
      // decimal place the leaf marker requires (`0d1e3` -> 10000/10^1).
      if (s < 1) {
        u *= pow10(1 - s)
        s = 1
      }
    }

    this.unscaled = u
    this.scale = s
  }


  // Exact numeric equality. Both operands are normalised by
  // construction, so equal values have equal fields -- this is value
  // comparison, never object identity (D2).
  equals(peer: Decimal): boolean {
    return this.unscaled === peer.unscaled && this.scale === peer.scale
  }


  // Exact ordering: -1, 0 or 1. Compares on a common scale, so no
  // float64 ever touches the operands.
  compare(peer: Decimal): number {
    let a = this.unscaled
    let b = peer.unscaled
    if (this.scale < peer.scale) {
      a *= pow10(peer.scale - this.scale)
    }
    else if (peer.scale < this.scale) {
      b *= pow10(this.scale - peer.scale)
    }
    return a < b ? -1 : a > b ? 1 : 0
  }


  negate(): Decimal {
    return new Decimal(-this.unscaled, this.scale)
  }


  add(peer: Decimal): Decimal {
    const scale = this.scale < peer.scale ? peer.scale : this.scale
    const a = this.unscaled * pow10(scale - this.scale)
    const b = peer.unscaled * pow10(scale - peer.scale)
    return new Decimal(a + b, scale)
  }


  multiply(peer: Decimal): Decimal {
    return new Decimal(this.unscaled * peer.unscaled, this.scale + peer.scale)
  }


  ceil(): Decimal {
    return new Decimal(this.intPart(1n), 0)
  }


  floor(): Decimal {
    return new Decimal(this.intPart(-1n), 0)
  }


  private intPart(dir: bigint): bigint {
    const p = pow10(this.scale)
    const q = this.unscaled / p
    const r = this.unscaled % p
    return 0n !== r && (0n < r) === (0n < dir) ? q + dir : q
  }


  isZero(): boolean {
    return 0n === this.unscaled
  }


  toString(): string {
    const neg = this.unscaled < 0n
    let d = (neg ? -this.unscaled : this.unscaled).toString()
    if (d.length <= this.scale) {
      d = '0'.repeat(this.scale - d.length + 1) + d
    }
    const cut = d.length - this.scale
    return (neg ? '-' : '') + d.slice(0, cut) + '.' + d.slice(cut)
  }


  // Canon rendering: the sign goes BEFORE the marker (`-0d1.5`), because
  // `0d-1.5` is not a literal this language accepts.
  canon(): string {
    const neg = this.unscaled < 0n
    return (neg ? '-0d' : '0d') + (neg ? this.negate() : this).toString()
  }


  static fromString(src: string): Decimal {
    const m = /^([-+]?)(?:0[dD])?([0-9]+)(?:\.([0-9]+))?(?:[eE]([-+]?[0-9]+))?$/
      .exec(src)
    if (null == m) {
      throw new Error('not-decimal: ' + src)
    }
    const frac = m[3] ?? ''
    const exp = null == m[4] ? 0 : Number(m[4])
    const scale = frac.length - exp

    if (overBudget(m[2].length + frac.length, scale)) {
      throw new Error('decimal-budget: ' + src)
    }

    const unscaled = BigInt(m[2] + frac)
    return new Decimal('-' === m[1] ? -unscaled : unscaled, scale)
  }
}


function decimalOverBudget(d: Decimal): boolean {
  return overBudget(coeffDigits(d.unscaled), d.scale)
}


const BIG_LITERAL_RE =
  /^0[dD]([0-9](?:_?[0-9])*)(?:\.([0-9](?:_?[0-9])*))?(?:[eE]([-+]?[0-9](?:_?[0-9])*))?/


type BigLiteral =
  { leaf: 'biginteger', int: bigint } |
  { leaf: 'bigdecimal', dec: Decimal } |
  { leaf: 'error', code: string }


function stripSep(s: string): string {
  return -1 === s.indexOf('_') ? s : s.replace(/_/g, '')
}


function readBigLiteral(m: RegExpExecArray | (string | undefined)[]): BigLiteral {
  const intd = stripSep(m[1] as string)
  const fracd = null == m[2] ? undefined : stripSep(m[2])
  const expd = null == m[3] ? undefined : stripSep(m[3])

  // Leaf by SOURCE, not by value.
  if (undefined === fracd && undefined === expd) {
    return { leaf: 'biginteger', int: BigInt(intd) }
  }

  const exp = undefined === expd ? 0 : Number(expd)
  const fracLen = undefined === fracd ? 0 : fracd.length
  const scale = fracLen - exp

  if (overBudget(intd.length + fracLen, scale)) {
    return { leaf: 'error', code: 'decimal_budget' }
  }

  return {
    leaf: 'bigdecimal',
    dec: new Decimal(BigInt(intd + (fracd ?? '')), scale),
  }
} /* node:coverage ignore next 12 */


export {
  BIG_LITERAL_RE,
  BigLiteral,
  DECIMAL_COEFFICIENT_BUDGET,
  DECIMAL_SCALE_BUDGET,
  Decimal,
  decimalOverBudget,
  readBigLiteral,
}
